lib/fog/compute/ecloud.rb

```diff
@@ -0,0 +1,822 @@
+          end
+          host_url = "#{params[:uri].scheme}://#{params[:uri].host}#{params[:uri].port ? ":#{params[:uri].port}" : ''}"
+
+          # Hash connections on the host_url ... There's nothing to say we won't get URI's that go to
+          # different hosts.
+          @connections[host_url] ||= Fog::XML::Connection.new(host_url, @persistent, @connection_options)
+
+          # Set headers to an empty hash if none are set.
+          headers = set_extra_headers_for(params) || set_extra_headers_for({})
+
+          # Make the request
+          options = {
+            :expects => (params[:expects] || 200),
+            :method  => params[:method] || 'GET',
+            :path    => params[:uri].path + "#{"?#{params[:uri].query}" if params[:uri].query}",
+            :headers => headers
+          }
+          unless params[:body].nil? || params[:body].empty?
+            options.merge!({:body => params[:body]})
+          end
+          response = @connections[host_url].request(options)
+          # Parse the response body into a hash
+          unless response.body.empty?
+            if params[:parse]
+              document = Fog::ToHashDocument.new
+              parser = Nokogiri::XML::SAX::PushParser.new(document)
+              parser << response.body
+              parser.finish
+
+              response.body = document.body
+            end
+          end
+
+          response
+        end
+
+        private
+
+        # if Authorization and x-tmrk-authorization are used, the x-tmrk-authorization takes precendence.
+        def set_extra_headers_for(params)
+          params[:headers] = {
+            'x-tmrk-version' => @version,
+            'Date'           => Time.now.utc.strftime("%a, %d %b %Y %H:%M:%S GMT"),
+          }.merge(params[:headers] || {})
+          if params[:method]=="POST" || params[:method]=="PUT"
+            params[:headers].merge!({"Content-Type" => 'application/xml'}) unless params[:headers]['Content-Type']
+            params[:headers].merge!({"Accept" => 'application/xml'})
+          end
+          unless params[:body].nil? || params[:body].empty?
+            params[:headers].merge!({"x-tmrk-contenthash" => "Sha256 #{Base64.encode64(Digest::SHA2.digest(params[:body].to_s)).chomp}"})
+          end
+          if @authentication_method == :basic_auth
+            params[:headers].merge!({'Authorization' => "Basic #{Base64.encode64(@username+":"+@password).delete("\r\n")}"})
+          elsif @authentication_method == :cloud_api_auth
+            signature = cloud_api_signature(params)
+            params[:headers].merge!({
+              "x-tmrk-authorization" => %{CloudApi AccessKey="#{@access_key}" SignatureType="HmacSha256" Signature="#{signature}"},
+              "Authorization" => %{CloudApi AccessKey="#{@access_key}" SignatureType="HmacSha256" Signature="#{signature}"}
+            })
+          end
+          params[:headers]
+        end
+
+        def cloud_api_signature(params)
```

Comment:
Assignment Branch Condition size for cloud_api_signature is too high. [18.97/15]